File: cpp/CodingTest/CodingTest/DoublyLinkedList.cpp
#include<iostream>

using namespace std;

template<class T>
struct DoublyLinkedListNode {
	T value;
    DoublyLinkedListNode<T>* prev;
	DoublyLinkedListNode<T>* next;
	DoublyLinkedListNode(T val) : value(val), prev(nullptr), next(nullptr) {}
};

template<class T>
class DoublyLinkedList {
private:
	DoublyLinkedListNode<T>* head;
	DoublyLinkedListNode<T>* tail;
public:
	unsigned int length;
	DoublyLinkedList() {
		head = nullptr;
		tail = nullptr;
		length = 0;
	}
	~DoublyLinkedList() {
		while (head) {
			pop_front();
		}
	}
	void push_front(T val) {
		DoublyLinkedListNode<T>* newHead = new DoublyLinkedListNode<T>(val);
		if (head == nullptr) tail = newHead;
		else {
			head->prev = newHead;
			newHead->next = head;
		}
		head = newHead;
		length++;
	}

	void push_back(T val) {
		DoublyLinkedListNode<T>* newTail = new DoublyLinkedListNode<T>(val);
		if (tail == nullptr) head = newTail;
		else {
			tail->next = newTail;
			newTail->prev = tail;
		}
		tail = newTail;
		this->length++;
	}
	void pop_front() {
		if (head == nullptr) return;
		if (head->next == nullptr) tail = nullptr;
		else {
			head->next->prev = nullptr;
			head = head->next;
		}
		if (length > 0) length--;
	}
	void pop_back() {
		if (tail == nullptr)  return;
		if (this->tail->prev == nullptr) this->head = nullptr;
		else {
			tail->prev->next = nullptr;
			tail = tail->prev;
		}
		if (length > 0) length--;
	}
	void remove(DoublyLinkedListNode<T>* node) {
		if (node.prev == nullptr && node.next == nullptr) {
			head = nullptr;
			tail = nullptr;
			length = 0;
			return;
		}
		if (node->prev == nullptr) {
			head = node->next;
			tail = node->prev;
		}
		else {
			node->prev->next->next = node->prev->next;
			node->prev->next->prev = node->prev;
		}
		if (length > 0) length--;
	}
};

int main() {
	//DoublyLinkedListNode<int>* a = new DoublyLinkedListNode<int>(1);
	//DoublyLinkedListNode<int>* b = new DoublyLinkedListNode<int>(10);
	//a->prev = b;
	//cout << a->prev->value << endl;
	DoublyLinkedList<int>* dll = new DoublyLinkedList<int>();
	dll->push_back(1);
	dll->push_back(2);
	dll->push_back(3);
	dll->pop_front();
	cout << dll->length << endl;

	return 0;
}

File: typescript/base_ds_algorithm/linked_list/doubly_linked_list.js
"use strict";
exports.__esModule = true;
var DoublyLinkedListNode = /** @class */ (function () {
    function DoublyLinkedListNode(value) {
        this.prev = null;
        this.value = value;
        this.next = null;
    }
    return DoublyLinkedListNode;
}());
//push_front, push_back, 
//pop_front, pop_back, 
//remove, 
//display_front, display_back
var DoublyLinkedList = /** @class */ (function () {
    function DoublyLinkedList() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    DoublyLinkedList.prototype.push_front = function (value) {
        var newHead = new DoublyLinkedListNode(value);
        if (this.head === null) {
            this.tail = newHead;
        }
        else {
            this.head.prev = newHead;
            newHead.next = this.head;
        }
        this.head = newHead;
        this.length++;
        return newHead;
    };
    DoublyLinkedList.prototype.push_back = function (value) {
        var newTail = new DoublyLinkedListNode(value);
        if (this.tail === null) {
            this.head = newTail;
        }
        else {
            this.tail.next = newTail;
            newTail.prev = this.tail;
        }
        this.tail = newTail;
        this.length++;
        return newTail;
    };
    DoublyLinkedList.prototype.pop_front = function () {
        if (this.head === null)
            return null;
        if (this.head.next === null) {
            this.tail = null;
        }
        else {
            this.head.next.prev = null;
            this.head = this.head.next;
        }
        if (this.length > 0)
            this.length--;
        return this.head;
    };
    DoublyLinkedList.prototype.pop_back = function () {
        if (this.tail === null)
            return null;
        if (this.tail.prev === null) {
            this.head = null;
        }
        else {
            this.tail.prev.next = null;
            this.tail = this.tail.prev;
        }
        if (this.length > 0)
            this.length--;
        return this.tail;
    };
    DoublyLinkedList.prototype.remove = function (node) {
        if (node.prev === null && node.next === null) {
            this.head = null;
            this.tail = null;
            this.length = 0;
            return node.value;
        }
        if (node.prev === null) {
            this.head = node.next;
            this.head.prev = null;
        }
        else {
            if (node.prev.next === null) {
                node.prev.next = null;
                this.tail = node.prev;
            }
            else {
                node.prev.next.next = node.prev.next;
                node.prev.next.prev = node.prev;
            }
        }
        if (this.length > 0)
            this.length--;
        return node.value;
    };
    return DoublyLinkedList;
}());
function runDoublyLinkedList() {
    var dll = new DoublyLinkedList();
    var testNode = new DoublyLinkedListNode(1);
    dll.push_front(1);
    console.log(dll);
    // dll.push_front(2);
    // dll.push_front(3);
    dll.remove(testNode);
    console.log(dll);
}
exports["default"] = runDoublyLinkedList;
